protostar/cheatable_starknet/cheaters/expect_events_controller.py

```diff
@@ -0,0 +1,142 @@
+
+if TYPE_CHECKING:
+    from protostar.testing import Hook
+    from protostar.cheatable_starknet.cheatables.cheatable_starknet_facade import (
+        CheatableStarknetFacade,
+    )
+    from protostar.cairo_testing import CairoTestExecutionState
+
+
+@dataclass
+class Event:
+    from_address: Address
+    key: Selector
+    data: Optional[CairoData] = None
+
+
+EventsExpectation = list[Event]
+
+
+@dataclass
+class EventMatching:
+    pass
+
+
+@dataclass
+class SkippedEventMatching(EventMatching):
+    emitted_event: Event
+
+
+@dataclass
+class AcceptedEventMatching(EventMatching):
+    emitted_event: Event
+    expected_event: Event
+
+
+@dataclass
+class FailedEventMatching(EventMatching):
+    expected_event: Event
+
+
+@dataclass
+class EventMatchingResult:
+    event_matchings: list[EventMatching]
+
+    @property
+    def failed_event_matchings(self):
+        return [
+            matching
+            for matching in self.event_matchings
+            if isinstance(matching, FailedEventMatching)
+        ]
+
+    @property
+    def should_be_accepted(self) -> bool:
+        return len(self.failed_event_matchings) == 0
+
+
+class ExpectEventsController:
+    def __init__(
+        self,
+        test_finish_hook: "Hook",
+        test_execution_state: "CairoTestExecutionState",
+        cheatable_starknet_facade: "CheatableStarknetFacade",
+    ) -> None:
+        self._test_execution_state = test_execution_state
+        self._cheatable_starknet_facade = cheatable_starknet_facade
+        self._test_finish_hook = test_finish_hook
+        self._test_finish_hook.on(self.compare_expected_and_actual_results)
+
+    def add_expectation(self, expected_events: list[Event]):
+        self._test_execution_state.event_expectations.append(expected_events)
+
+    def compare_expected_and_actual_results(self):
+        for expected_events in self._test_execution_state.event_expectations:
+            matching_result = match_events(
+                expected_events=expected_events,
+                emitted_events=self._cheatable_starknet_facade.get_emitted_events(),
+            )
+            if not matching_result.should_be_accepted:
+                raise ExpectEventsMismatchException(matching_result)
+
+
+def match_events(
+    expected_events: list[Event], emitted_events: list[Event]
+) -> EventMatchingResult:
+    assert len(expected_events) > 0
+    expected_events_queue = deque(expected_events)
+    event_matchings: list[EventMatching] = []
+    for emitted_event in emitted_events:
+        try:
+            event_matchings.append(
+                match_expected_and_emitted_event(
+                    emitted_event=emitted_event,
+                    expected_event=expected_events_queue[0],
+                    on_accepted=expected_events_queue.popleft,
+                )
+            )
+        except IndexError:
+            event_matchings.append(SkippedEventMatching(emitted_event=emitted_event))
```

Comment:
I'd rather see a check for a list's item but it's a NIT, for me it just would be cleaner.

Reply:
IndexError might be raised from `match_expected_and_emitted_event` method, it's not a nit @karol-bisztyga  :)

Reply:
[simplify match_events](https://github.com/software-mansion/protostar/pull/1412/commits/0a93f31859a66511d69a776cc6b4edb15bb0ab46)